File: backend/app/schemas/request.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.models import RequestPriority, RequestStatus


class RequestCreate(BaseModel):
    title: str = Field(min_length=3, max_length=120)
    description: str | None = Field(default=None, max_length=1000)
    priority: RequestPriority = RequestPriority.normal


class RequestStatusUpdate(BaseModel):
    status: RequestStatus


class RequestResponse(BaseModel):
    id: int
    title: str
    description: str | None
    status: RequestStatus
    priority: RequestPriority
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class RequestListResponse(BaseModel):
    items: list[RequestResponse]
    total: int
    page: int
    page_size: int
    pages: int


SortField = Literal["created_at", "priority"]
SortOrder = Literal["asc", "desc"]

File: frontend/src/types/request.ts
export type RequestStatus = "new" | "in_progress" | "done";
export type RequestPriority = "low" | "normal" | "high";
export type SortField = "created_at" | "priority";
export type SortOrder = "asc" | "desc";

export interface RequestItem {
  id: number;
  title: string;
  description: string | null;
  status: RequestStatus;
  priority: RequestPriority;
  created_at: string;
  updated_at: string;
}

export interface RequestListResponse {
  items: RequestItem[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

export interface RequestCreatePayload {
  title: string;
  description?: string | null;
  priority: RequestPriority;
}

export interface RequestFilters {
  status?: RequestStatus;
  priority?: RequestPriority;
  search?: string;
  sort_by: SortField;
  sort_order: SortOrder;
  page: number;
  page_size: number;
}

export interface ApiError {
  detail: string;
}
